### DISCUSSION TITLE
battery parameters

### DISCUSSION THREAD

**Discussion by madtatu-development** [KG-Development/SmartBMSUtility#12] [Q&A]
Hi,

I was so excited when I found Smart BMS utility iOS app, as after purchasing my BMS with BLE module, I thought there would be only Android apps out there. But I got lost, and I am sure it is 100% me, being dummy, so I would like to ask your kind help in my battery configuration question:

**Battery Specs:**
- 10S17P Battery Pack (custom built by myself)
- Cells type: LG INR18650-MJ1
- So in total: 59.5Ah, 170A, 36V
- BMS type: LH-SP15S008 with BLE module and 2 temp sensor (all attached)

**Single cell specs:**
- Voltage charge max.: 4.20 V
- Voltage, nominal: 3.65 V
- Voltage, discharge end: 2.50 V
- Capacity, max.: 3500.00 mAh
- Constant charge current, max.: 3.40 A
- Charge constant standard: 1.7 A
- Discharge standard: 680 mA
- Max. continuous discharge current: 10.00 A
- Peak discharge theoretic, 4 sec: 17.5 A
- C-rate (charge, max.): 0.97 C
- C-rate (discharge, max.): 2.86 C

**Status:**
- Battery is fully charged (each separate cells were measured too for countercheck)
- Smart BMS utility can find the BMS and connect to it perfectly. 
- The basic data its read from BMS are:

![basic_1](https://user-images.githubusercontent.com/73760229/121816997-1cbd1a00-cc7f-11eb-8e8e-9317c7136499.jpg)
![basic_2](https://user-images.githubusercontent.com/73760229/121817009-25155500-cc7f-11eb-930a-4cc41767fc07.jpg)

- The configuration data:

![config_1](https://user-images.githubusercontent.com/73760229/121817665-02853b00-cc83-11eb-9b04-f5cbe3219604.jpg)
![config_2](https://user-images.githubusercontent.com/73760229/121817670-0618c200-cc83-11eb-8ed4-4a714e75fd5b.jpg)
![config_3](https://user-images.githubusercontent.com/73760229/121817672-087b1c00-cc83-11eb-8836-c5a1630abe17.jpg)
![config_4](https://user-images.githubusercontent.com/73760229/121817675-0a44df80-cc83-11eb-9111-041b9e9cd801.jpg)
![config_5](https://user-images.githubusercontent.com/73760229/121817677-0b760c80-cc83-11eb-98ba-07e65102ea80.jpg)
![config_6](https://user-images.githubusercontent.com/73760229/121817679-0d3fd000-cc83-11eb-8fdb-7c2e2132d3d9.jpg)
![config_7](https://user-images.githubusercontent.com/73760229/121817682-0f099380-cc83-11eb-8479-db5ff1cc89f3.jpg)

**---- The PROBLEM ----**
- I am UNABLE to TURN ON the "Discharge" button, because it triggers the following error: "Warning battery Undervoltage!"

**---- QUESTION ----**
- Triple checked setup, cables, cells, etc, and confirmed to be perfect.
- So, I assume, this problem comes form the fact, that the BMS have to be configured.
**- I searched a lot on Google and also read the Overkill Solar doc, but still as I am not 100% sure, to be "better safe then sorry", may I ask your kind help what's the best value to write into these config fields (based on specs above) to ensure that I am able to use my battery ? :**

Number of cells:  

Total battery capacity:
Total cycle capacity: 
Cell selfdischarge rate:
Cell full voltage:
Cell 80% charge voltage:
Cell 60% charge voltage:
Cell 40% charge voltage:
Cell 20% charge voltage:
Cell empty voltage:

Balancing start voltage:
Cell delta target:
Enable balancing : ON/OFF
Balancing only whilst charging: ON/OFF

Cell overvoltage
--Trigger:
--Release:
--Delay:

Cell undervoltage
--Trigger:
--Release:
--Delay:

Pack overvoltage:
--Trigger:
--Release:
--Delay:

Pack undervoltage
--Trigger:
--Release:
--Delay:

Charge overtemperature
--Trigger:
--Release:
--Delay:

Charge undertemperature
--Trigger:
--Release:
--Delay:

Discharge overtemperature
--Trigger:
--Release:
--Delay:

Discharge undertemperature
--Trigger:
--Release:
--Delay:

**I know I am asking a lot of data help here, which should be logically come from specs data, but I promise, I am not asking this help because I am lazy or such. I am just desperately need a professional help to confirm a good setup to fully understand all fields.

Many many many many many many BIG thanks for your time and help,**

Andras

**Comment by NeariX67** [comment]
i really really appreciate your detailed list. It makes helping people like you much much easier.

So you have a 10S battery pack. So the lower battery voltage (assuming you use 3000mV instead of the cell cutoff voltage of 2500mV) would be at `10*3000mV = 30000mV` and your full battery voltage should be `10*4200mV`, so 42000mV.
You currently have it set to 42000mV/64000mV so **1.** in the wrong order and **2.** wrong values

This should resolve your issues:
Pack overvoltage:
--Trigger: 42500mV (just a little bit over the calculated 42000mV)
--Release: 42000mV (just right what we calculated)
--Delay: (whatever you want - or keep default)

Pack undervoltage
--Trigger: 29500mV (a little bit under the calculated 30000mV, you might go as low as 25000mV)
--Release: 30000mV
--Delay: (whatever you want - or keep default)

If you experience any other issues i'll gladly help you.

**Reply by madtatu-development** [reply]
THANK YOU so MUCH for your kind and quick reply !!

Sharing all possible details is the minimum when asking somebody's help as you are spending your valuable time on it, so if you need guessing or need to figure out the details from an unstructured chaos, that would waste your time, what I would never do, I know by experience (when I am on the other side) how bad and non-respectful it can be . 

All clear on what you had kindly written (and again BIB BIG BIG thanks), just before I write the settings into BMS, may you allow me to ask your kind help on the correct values of the rest of the parameters below (if you recommend to update those too, for a perfect setup)? 

I had entered a value with an explanation why, in hope that would make it clear if I assume something wrong:

- Total battery capacity: 59 500mAh (as this is the TOTAL battery capacity of a 10S17P with 3500 mAh cells)
- Total cycle capacity: 47 600mAh as 80% of (as it seems, that without several charging circles and measures, this is a "typical %")
- Cell selfdischarge rate: 0.2% (I could not found anything in cell factory specs sheet, so I left it as the default.)
- Cell full voltage: 42 000mV (assuming, that "cell" title here refers to 1 serial row and not single cell unit max charging voltage)
- Cell 80% charge voltage: 33 600mV (80% of 4200mV)
- Cell 60% charge voltage: 25 200mV (60% of 4200mV)
- Cell 40% charge voltage: 16 800mV (40% of 4200mV)
- Cell 20% charge voltage: 8 400mV  (20% of 4200mV)
- Cell empty voltage: 2 500mV (based on the specs voltage discharge end value)

- Balancing start voltage: 3 650 mV (assuming, that this refers to the nominal voltage?)
- Cell delta target: 50 (I left it as default, because I could not found any digestible info on this)
- Enable balancing : ON (assuming, this is always a good thing, based on searching)
- Balancing only whilst charging: OFF (again, assuming this is the best way to go)

Finally, may I ask to confirm, that if I make a donation to @nearix846, you will got it ?

Many many many thanks again for your time and help,

Andras

**Reply by NeariX67** [reply]
Okay so:

Total battery capacity seems right
Total cycle capacity can be set, but will be measured by the bms (after 1-3 full charge and discharge cycles) so setting the correct values doesn't really matter here
Cell selfdischarge is right
Cell Full voltage is 4200mV not 42000mV (my app should warn you about that)
Cell 80-20% are best as they were by default (4000,3800,3600,3300)
Cell empty voltage should be either 3000mV if you want long lifetime out of your batteries or 2500mV if you want every mAh out of it (you may need to adjust 80-20% voltages again)
Balancing start voltage can be set to 4000, 3600 or even 3300. Doesn't really matter. I recommend 4000-4100mV.
Cell voltage target of 50mV seems a bit high, bms manufacturer recommends 15-30mV
Balancing ON is a must!

To the "balancing only whilst charging":
The bms has a little annoying feature. It either balances when its charging (not after it's done!) or when it is not charging! Either of these options are not the perfect solution. The perfect solution would be a combination of both but we can't modify that one.

Donations to my linkes paypal.me will get to me 100%.
Appreciate your donation.

Have a good one

**Reply by madtatu-development** [reply]
**You ROCK !

I can not express how much I appreciate your patience, and detailed help. From answers like yours, possible to learn a lot, so THANK YOU! THANK YOU! THANK YOU!**

I set it up, written it to BMS, and worked like charm, the only 2 things I noticed in my quick test are:

1) There is a possible "bug" caused by a combination of steps, which RESET the BMS configuration to default (so have to re-enter all value and write it again to BMS). I ran into it 2-3 times, but I could not solidly reproduced it yet for you, yet. It seems it is a combination/sequence of charge/discharge button push + BLE + config read button push.  Anyway, as soon as I am able to catch it, I send you a detailed report which based on, you can hunt it down. 

2) The Overview screen main battery level gauge is ALWAYS zero:

![level](https://user-images.githubusercontent.com/73760229/121824996-480c2d00-ccb0-11eb-9776-fd9aded78bc8.jpg)

(The only time it shown the correct value, the EXACT moment, when BMS config got reset, described in step nr. 1.

I know it is not enough information yet, but by any chance do you have an educated guess what could cause this problem ?

Finally, thank you for the paypal confirmation, I do the donation, as great app, great support!

Many many many many thanks,

Andras

**Reply by NeariX67** [reply]
Glad that everything worked like it should. Issue 1 sounds interesting but I'm currently not sure if that issue is on my side.
Issue 2 is definitely not an issue on my side. Charge your battery completely, the bms will notice that your battery is full and it will set the state of charge to 100%

**Reply by madtatu-development** [reply]
Thanks for your reply! 

May I ask:
- I am using a li ion charger, which automatically "detect" if battery is fully charged. For some reason it always stop charging this battery around 92% perfect. Since theoretically the most healthy for the battery is to keep it around 80-85%, it never bothered me. 
**BUT, does it mean, if it does not reach what BMS believes as 100%, it would never trigger the level gauge indicator ?**

**Reply by NeariX67** [reply]
This could indeed be a problem. The way I would handle your situation would be the following:

Assuming that your charger stops at 40.9V instead of 42.00V (taken from your screenshots) you could set your "battery full voltage" parameter to like 4090mV and your Cell Overvoltage protection to 4090mV and the Pack overvoltage trigger to 40900mV. That way your battery will always stop charging at that voltage (instead of stopping because of your charger).

**Reply by madtatu-development** [reply]
Thank you for your quick reply again!

**So, what I did, step by step:**

1) I had attached the charger again after couple of hours to se if that triggers any additional charge, and it did. Not much, but some. After this, I was waiting again for couple hours and attached the charger again, but that did not triggered any further charging, so the final number (41.52V) looks as follows:

![max_charge](https://user-images.githubusercontent.com/73760229/121896963-2cda0580-cd22-11eb-9e96-d3f9e58d32a1.png)

2) After this I followed your advise, with updated numbers (according to the pattern you advised), so set BMS as follows:
Capacity Settings
- Cell full voltage: 4152mV

Protection settings
- Cell overvoltage
Trigger: 4152mV
Release: 4052mV

- Pack overvoltage
Trigger: 41520mV
Release: 41020mV

3) Restarted the app, and connected to the BMS again, and:
- GOOD NEWS: battery level gauge indicator started to work !
- **BAD NEWS:** I received the following 2 warnings:
"Cell over voltage"
"Battery over voltage"

4) I tried:
- A) to set a higher value for cell overvoltage (t:4190,r:4090) compared to "Highest voltage" value at Overview screen
- B) to set a lower value for cell overvoltage (t:4100,r:4000) compared to "Highest voltage" value at Overview screen

BUT I couldn't get rid of the warnings. (Though battery discharge seems to work perfectly, what was the main target.)

May I ask if you kind advise, what am I doing wrong ?

VERY BIG thank you in advance!!

**Reply by NeariX67** [reply]
I'd suggest some minor adjustments: Cell Full voltage to around 4175mV.

Cell overvoltage should also be set to trigger at 4175mV and release 4140mV
Pack overvoltage can stay at 42000mV and 41800mV without any issues.

You can check for errors and warning when you are connected to your battery under "more" -> "Protection Status".

Cell overvoltage and battery overvoltage are just a warning. Nothing serious. It's even good and desired to receive such warnings.

**Reply by madtatu-development** [reply]
Sorry for the slow reply, but I had the opportunity to use the battery in a longer period for testing, so I wanted to wait till the end to be able to reply with more comprehensively:

**First of all, I did all minor adjustments you suggested. So the final setup is:**

![capacity](https://user-images.githubusercontent.com/73760229/122561711-7a1be700-d042-11eb-802e-57acd07ed88d.jpg)
![balancing](https://user-images.githubusercontent.com/73760229/122561721-7daf6e00-d042-11eb-8885-b88ef6461997.jpg)
![protection](https://user-images.githubusercontent.com/73760229/122561732-7f793180-d042-11eb-9e69-2756511e9797.jpg)

I also checked the errors as you sugegsted:

![warnings](https://user-images.githubusercontent.com/73760229/122561814-96b81f00-d042-11eb-94f5-5c88feecc8cc.jpg)

**After 1 hour of usage, when the battery level gauge went down to 99% both "Cell over voltage" and "Battery over voltage" disappeared. Exactly at this point:**

![used](https://user-images.githubusercontent.com/73760229/122561970-c6672700-d042-11eb-8897-d6980ed4a657.png)

_App conclusions based on regular usage:_
- Superior, I love it! Not to mention the support which is the best I ever received!!
- If you allow me to dream, what would be so user friendly to have:

**1) Possibility in settings to select a default BMS, which could load automatically at startup, bypassing this screen (as when you have only 1 BMS, this is a useless step):**

![IMG_2802](https://user-images.githubusercontent.com/73760229/122565648-029c8680-d047-11eb-9b5e-b710afcb464a.PNG)

**2) Rename either these onces below from "Cell" to "Cell Series"** 

![IMG_2803](https://user-images.githubusercontent.com/73760229/122565918-45f6f500-d047-11eb-877a-56e2527301ed.jpg)

**OR**

**These ones from "Cell" to "Single Cell"**

![IMG_2804](https://user-images.githubusercontent.com/73760229/122565995-59a25b80-d047-11eb-8189-abd24a092b3b.jpg)
![IMG_2805](https://user-images.githubusercontent.com/73760229/122565973-54451100-d047-11eb-8d74-fd057b283c46.jpg)

BECAUSE the world "Cell" used on both cases, though one means a series, the other refers to a single Cell is totally confusing, at least for such idiots like me :)

3) Fixing strange buttons behaviors, happens ONLY at FRESH app start. The 2 buttons acts reverse. If you push charge, discharge act. If you push discharge, charge acts. BUT only after a fresh app start. Please have a look at the videos below, I recorded for you to be able to replicate the problem:

https://user-images.githubusercontent.com/73760229/122566651-13013100-d048-11eb-90f7-e39ccd216f40.MP4

https://user-images.githubusercontent.com/73760229/122566144-822a5580-d047-11eb-8ba1-13b413f8f842.MP4

**4) Would be awesome to have a possibility in settings to hide not serious errors such as : "Cell over voltage", "Battery over voltage" completely or at least after once shown, as they can be very disturbing, and the since the error screen is there too, an option for regular users to disable it in settings would be very handy.** 

**5) Fixing the overlaying error notifications. If there i snore than one error show on the top, the second one covers the first one.** 

Finally, I am trying to make you a donation what i promised, but I ran into a problem:

![{76580EE5-DB63-4B6B-B43C-146B928D6B70}](https://user-images.githubusercontent.com/73760229/122567420-edc0f280-d048-11eb-99de-69dfdb5401d4.png)

Many thanks,

Andras

**Reply by NeariX67** [reply]
Thank you for your honest feedback. I'd happily implement your suggested "dreams".

1) Default BMS sounds like a good idea. I didn't had time in the last weeks to maintain the app so I will make some crash fixes and update the app. If thats done I'll start working on a complete new frontend. I'll happily implement your default bms suggestion as well as 4) and 5).

5. might be difficult as it is nothing I've coded and just used a public project that does these notifications. I'll hope i find something different that fixes that.

2) 

> BECAUSE the world "Cell" used on both cases, though one means a series, the other refers to a single Cell is totally confusing, at least for such idiots like me :)

Cell on the overview page indeed refers to "cell series" but it would make the text way bigger than necessary which could also look very crowded on smaller devices like the iPhone 5S. I'm happy like it currently is.

3) I've noticed this too. I already took a short look into this issue a while back but couldn't find the cause. I'll adress that in version 2.0.

<hr>

I don't what platform you are trying to donate me on, but the quickest and easiest way is with my [paypal.me](https://www.paypal.com/paypalme/nearix846) link


Btw: Warning are just warning. If you receive and warning about overvoltage or overtemperature or such. That is good. No need to worry or adjust the configuration perfectly just to avoid warnings.

**Reply by madtatu-development** [reply]
Thank you for your quick reply and your positive approach!

1) 4) 5) BIG thanks!

3) I guess viewcontrolers got mixed up. Easy fix suggestion:
- What if you replace this type of button by the iOS standard on/off switch ?

2) May I suggest a "golden middle way"?
- What if you simple rename the tite only form "CELLVOLTAGES" to -> "CELL-SERIES VOLTAGES" or "CELL-S VOLTAGES" which would both fit perfectly on all devices. So an elegant and easy way. 

Donation) Actually, the screenshot was made in paypal/send money screen... BUT the paypalme link you sent worked perfectly, so donation made (Fitexport Kft in notes). :)

May I aks whats the plans for v2? Seems you plan a complete revamp.

**Reply by NeariX67** [reply]
Appreciate your donation. People like you help the open-source community a lot! Love it!

I'm gonna work a little bit on the app today and might be able to implement your auto-connect request already.
3. just got fixed (buttons), I didn't update local values and they used the default "true" for both charging and discharging so it would toggle the state no matter what it is actually set to.

"Cell-series voltages" will be updated in v1.2.2 :)


Yes, I indeed plan a complete overhaul of the app. I am currently using "Storyboards" (don't know if you know that this?) which is quite old and not the way apple want's apps to be made of anymore. The new SwiftUI interfacing method will make the app way more attractive and fitting to other apps. But redesigning the complete app will probably take several weeks.

My todo for today is implementing your autoconnect feature and then I'll take a look into the app crashes that only occur on iPads.

**Reply by madtatu-development** [reply]
Thanks to You mate!! You know, its only possible because you are also awesome! I ran into soooo many times, the developpers prefers "take it or leave it" style, and when I ask a question which too dummy for them, regardless how polite I am they simple
ignore it. Not you! You were ready to help, and which builds trust and confidence! So thank You!!

3) I see! Great you found it, sometimes these little things could drive you crazy. Awesome!!!

Cell-Series Voltages) BIG thank you!!! It wil make it more logical for dummies like me. 

v2.0) Yeah, I know what you mean, I am a lot into ios app develoment. You are 100% right thats quite old, but I also know that to move to SwiftUI interfacing, can be in the pain in the ass, basically a new app, could take lots of time and energy. In case you need a Testflight tester , juts let me know and I will send my ios account email in private. 

Autoconnect) Super great news!!!! BIG thanks!!!

Few more ideas (though I know it is a bit crazy one) if you allow me: 

You know what would be super cool?

6)
- As my experience shows the overvoltage errors dissapears after a certain point, which means also, that it can be detected. 
- What if the app would have a semi-automatic option (could be enabled in settings, so everybody could decide if prefer it ot not) which would automatically change (or offer the change) to update overvoltage protection values, based on the the detected "treshold".

7) 
- Possibility to save the BMS setup with custom name, which would be stored in iCloud after its saved. 
- This way, even when BMS got reset, it would be easy to re-load it. 
- (I am still trying to catch the error I mentioned you earlier, when on certain charge/discharge button press combination, the BMS got overwritten with default values. Again, unlikely this is your app bug, BUT still a save/load option would help to manage such situation. Not to mention other scenarios it could help with as well)

8)
- Would be so great to have an option to load configuration automatically at startup. Could be very handy. 

9)
- Since you have GPS implemented, and app can run in background. 
- A kind of "Fuel Consuption" feature. The idea came from the fact, that I am using my battery on a custom e-bike. 
- So a kind of "dashboard" with the power level gauge AND a "Distance left" bar and km indication, calculated using the KMs taken and the battery value used for this KMs (so only in motion, not when stopped). 

Finally, I read in an other discussion that you were aiming for precise speed fromGPS, which is pretty hard in iphone based on my experience. But if you actually read the distance from apple map kit combined with time, that would give you pretty good value. 

Sorry for writing so much, I am just enthusiastic :)

**Reply by NeariX67** [reply]
Sounds like a good todo for 2.0!

**Reply by madtatu-development** [reply]
:)))))))

**Reply by NeariX67** [reply]
v1.2.2 is out now :)

**Reply by madtatu-development** [reply]
Yeah, awesome!!

Can confirm:
- Buttons are working correctly
- Auto connect working correctly 

Small UI issue:
- The very First time when go to "more", the autoconnect checkbox was checked, but the autoconnect feature did not worked I had to uncheck and check again to make it work. Solution: if this feature is disabled as default, than the checkbox default value should be disabled.  BUT again, this is just cosmetical issue. 

BIG thanks!!!!!

**Reply by madtatu-development** [reply]
Hello Sir, May I ask if you received my post above ?

**Reply by NeariX67** [reply]
If you mean your post from 4 days ago, yes, otherwise no

**Reply by madtatu-development** [reply]
Yeah I mean that. :) Sorry for asking, just you usually respond pretty fast, so I just wanted to be sure.

**Reply by NeariX67** [reply]
Yeah, i just didn't want to publish a new update just for this very very small issue. But i've got a message from another user that found another small issue that might be worth an update. So i'd fix that if the other issue is actually an issue.

**Reply by madtatu-development** [reply]
Absolutely Sir, no rush at all. Evenmore I should be more precise and use the "Issues" tab here on Github, instead of the discussion part. You are awesome, I oove your app!
